modules/ip/control/route.c

```diff
@@ -301,6 +301,76 @@ static void route4_fini(struct event_base *) {
 	vrf_fibs = NULL;
 }
 
+void ip4_route_cleanup(uint16_t vrf_id, struct nexthop *nh) {
+	struct rte_rib_node *rn = NULL;
+	struct rte_rib *rib;
+	uint32_t local_ip;
+	uint8_t prefixlen;
+	uint64_t nh_idx;
+	uint32_t mask;
+	uint32_t ip;
+
+	local_ip = nh->ip;
+	mask = rte_be_to_cpu_32(~(0xffffffff >> nh->prefixlen));
+
+	rib = rte_fib_get_rib(get_fib(vrf_id));
+	while ((rn = rte_rib_get_nxt(rib, 0, 0, rn, RTE_RIB_GET_NXT_ALL)) != NULL) {
+		rte_rib_get_nh(rn, &nh_idx);
+		nh = ip4_nexthop_get(nh_idx);
+
+		if (nh && (nh->ip & mask) == (local_ip & mask)) {
+			rte_rib_get_ip(rn, &ip);
+			rte_rib_get_depth(rn, &prefixlen);
+			ip = rte_be_to_cpu_32(ip);
+
+			LOG(DEBUG,
+			    "%s ip4_route_delete %d.%d.%d.%d/%d via %d.%d.%d.%d",
+			    __func__,
+			    ((uint8_t *)&ip)[0],
+			    ((uint8_t *)&ip)[1],
+			    ((uint8_t *)&ip)[2],
+			    ((uint8_t *)&ip)[3],
+			    prefixlen,
+			    ((uint8_t *)&nh->ip)[0],
+			    ((uint8_t *)&nh->ip)[1],
+			    ((uint8_t *)&nh->ip)[2],
+			    ((uint8_t *)&nh->ip)[3]);
```

Comment:
Logging IP and ethernet addresses seem quite frequent. We need wrappers around `rte_log` and `snprintf` to reduce boilerplate. E.g.:

```diff
-LOG(DEBUG,
-    "%s ip4_route_delete %d.%d.%d.%d/%d via %d.%d.%d.%d",
-    __func__,
-    ((uint8_t *)&ip)[0],
-    ((uint8_t *)&ip)[1],
-    ((uint8_t *)&ip)[2],
-    ((uint8_t *)&ip)[3],
-    prefixlen,
-    ((uint8_t *)&nh->ip)[0],
-    ((uint8_t *)&nh->ip)[1],
-    ((uint8_t *)&nh->ip)[2],
-    ((uint8_t *)&nh->ip)[3]);
+ LOG(DEBUG, "%4/%hhu via %4", &ip, &nh->ip);
```

Or:

```diff
-scols_line_sprintf(line, 2, ETH_ADDR_FMT, ETH_BYTES_SPLIT(nh->mac.bytes));
+scols_line_sprintf(line, 2, "%2", &nh->mac));
```

What do you think?